source/common/upstream/load_balancer_impl.cc

```diff
@@ -1299,19 +1299,80 @@ HostConstSharedPtr LeastRequestLoadBalancer::unweightedHostPick(const HostVector
                                                                 const HostsSource&) {
   HostSharedPtr candidate_host = nullptr;
 
+  switch (selection_method_) {
+    case envoy::extensions::load_balancing_policies::least_request::v3::LeastRequest::FULL_SCAN:
+      candidate_host = unweightedHostPickFullScan(hosts_to_use);
+      break;
+    case envoy::extensions::load_balancing_policies::least_request::v3::LeastRequest::N_CHOICES:
+      candidate_host = unweightedHostPickNChoices(hosts_to_use);
+      break;
+    default:
+      IS_ENVOY_BUG("unknown selection method specified for least request load balancer");
+  }
+
+  return candidate_host;
+}
+
+HostSharedPtr LeastRequestLoadBalancer::unweightedHostPickFullScan(const HostVector& hosts_to_use) {
+  HostSharedPtr candidate_host = nullptr;
+
+  uint32_t num_hosts_known_tied_for_least = 0;
+
+  const uint32_t num_hosts = hosts_to_use.size();
+
+  for (uint32_t i = 0; i < num_hosts; ++i) {
+    const HostSharedPtr& sampled_host = hosts_to_use[i];
+
+    if (candidate_host == nullptr) {
+      // Make a first choice to start the comparisons.
+      num_hosts_known_tied_for_least = 1;
+      candidate_host = sampled_host;
+      continue;
+    }
+
+    const auto candidate_active_rq = candidate_host->stats().rq_active_.value();
+    const auto sampled_active_rq = sampled_host->stats().rq_active_.value();
+
+    if (sampled_active_rq < candidate_active_rq) {
+      // Reset the count of known tied hosts.
+      num_hosts_known_tied_for_least = 1;
+      candidate_host = sampled_host;
+    } else if (sampled_active_rq == candidate_active_rq) {
+      ++num_hosts_known_tied_for_least;
+
+      // Use reservoir sampling to select 1 unique sample from the total number of hosts N
+      // that will tie for least requests after processing the full hosts array.
+      //
+      // Upon each new tie encountered, replace candidate_host with sampled_host
+      // 1 / num_hosts_known_tied_for_least percent of the time.
```

Comment:
```suggestion
      // with probability (1 / num_hosts_known_tied_for_least percent).
```

Reply:
I know this is a bit pedantic, but given that we're being fancy here it's prudent to avoid misunderstandings where we can.